network_simulator/statistics.py

```diff
@@ -4,6 +4,200 @@
+        self.flow_stats = flow_stats
+        self.host_stats = host_stats
+
+    def get_link_stats(self, link):
+        """
+        Check whether a link has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Link link: link to be determined.
+        :return :class:`.LinkStats`.
+        """
+        # Create a tuple of the pair and check if it exists in the dict.
+        tup = (link.end_1_addr, link.end_2_addr)
+
+        if tup not in self.link_stats:
+            self.link_stats[tup] = LinkStats()
+        return self.link_stats[tup]
+
+    def get_flow_stats(self, flow):
+        """
+        Check whether a flow has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Flow flow: flow to be determined.
+        :return :class:`.FlowStats`.
+        """
+        if flow.flow_id not in self.flow_stats:
+            self.flow_stats[flow.flow_id] = FlowStats()
+        return self.flow_stats[flow.flow_id]
+
+    def get_host_stats(self, host):
+        """
+        Check whether a host has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Host host: host to be determined.
+        :return :class:`.HostStats`.
+        """
+        if host.address not in self.host_stats:
+            self.host_stats[host.address] = HostStats()
+        return self.host_stats[host.address]
+
+    def link_buffer_occ_change(self, link, curr_time):
+        """
+        Update the buffer size as a link gains or loses a queue element.
+
+        :param Link link: link that has a change of buffer size.
+        """
+        stats = self.get_link_stats(link)
+        # TODO(sharon): Check with Cody if qsize() would work in this case.
+        buffer_occ_packets = link.buffer.qsize()
+        stats.buffer_occupancy.append((curr_time, buffer_occ_packets))
+
+    def link_packet_loss(self, link, curr_time):
+        """
+        Record a packet loss of a link.
+
+        :param Link link: link of action.
+        :param float curr_time: simulation time of loss.
+        """
+        stats = self.get_link_stats(link)
+        stats.packet_loss_times.append(curr_time)
+
+    def link_packet_transmitted(self, link, packet, curr_time):
+        """
+        Record a packet transmission of a link.
+
+        :param Link link: link of action.
+        :param int packet_size: the size in bits of packet transmitting.
+        :param float curr_time: simulation time of transmission.
+        """
+        stats = self.get_link_stats(link)
+        stats.packet_transmit_times.append((curr_time, packet_size))
+
+    def flow_packet_sent(self, flow, curr_time):
```

Comment:
Might be easier to pass in the DataPacket itself here, and just read off its start_time_sec, instead of passing in curr_time.